[Kafka] Change Kafka installation - multi namespace support

shay79il opened the pull request:
Migrating bitnami kafka to Strimzi Kafka operator

Add Strimzi Kafka operator configuration 
Update values for Kafka deployment 

[JIRA](https://iguazio.atlassian.net/browse/CEML-492)

shay79il committed:
[Kafka] Change Kafka installation - multi namespace support

Add Strimzi Kafka operator configuration and update values for Kafka deployment
[JIRA](https://iguazio.atlassian.net/browse/CEML-492)

GiladShapira94 commented:
Looks good, can you edit the `admin` and the `non-admin` values files with the values that need to be use.
it does not need to support ingress and user that use `NodePort` will need to change the ports manually

GiladShapira94 commented on charts/mlrun-ce/templates/kafka/kafka-nodepool.yaml:
```diff
@@ -0,0 +1,30 @@
+kind: KafkaNodePool
+metadata:
+  name: {{ .Values.kafka.name }}-pool
+  namespace: {{ .Values.kafka.namespace | default .Release.Namespace }}
```
@shay79il one small things, can you please specify what is the image and the image tag we have been using in the values.yaml file

shay79il commented on charts/mlrun-ce/templates/kafka/kafka-nodepool.yaml:
```diff
@@ -0,0 +1,30 @@
+  storage:
+    type: {{ .Values.kafka.storage.type }}
+    size: {{ .Values.kafka.storage.size }}
+    {{- if .Values.kafka.storage.class }}
+    class: {{ .Values.kafka.storage.class }}
+    {{- end }}
```
Copilot's comment is wrong here. 
My code already does exactly what it suggested - it removes the class field when it's empty.

shay79il commented on charts/mlrun-ce/templates/kafka/kafka-cluster.yaml:
```diff
@@ -0,0 +1,31 @@
+    {{- end }}
+  {{- if gt (.Values.kafka.zookeeper.replicas | int) 0 }}
+  zookeeper:
+    replicas: {{ .Values.kafka.zookeeper.replicas }}
```
duplicate lines!

shay79il reviewed (commented):
.

shay79il commented on charts/mlrun-ce/templates/kafka/kafka-bootstrap-alias.yaml:
```diff
@@ -0,0 +1,24 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: {{ .Values.kafka.bootstrapAlias.name | default "kafka-stream" }}
```
Done

shay79il commented on charts/mlrun-ce/values.yaml:
```diff
@@ -468,37 +465,69 @@ tdengine:
+  # Bootstrap service alias configuration
+  bootstrapAlias:
+    # Create a service alias for simpler Kafka bootstrap server name
+    # When enabled, creates: {aliasName}.{namespace}.svc.cluster.local:9092
```
Done

shay79il commented:
# MLRun CE Install Guide

### Prerequisites 

### 1) Install ingress-nginx (optional)

If you want ingress URLs to work on k8s

```bash
kubectl apply -f https://raw.githubusercontent.com/kubernetes/ingress-nginx/controller-v1.11.3/deploy/static/provider/cloud/deploy.yaml
kubectl rollout status -n ingress-nginx deployment/ingress-nginx-controller --timeout=2m
```

### 2) Add /etc/hosts entries (per namespace)

Add these for EACH namespace you deploy (example for `mlrun` + `docker-desktop`):

```bash
sudo /bin/sh -c 'printf "\n# mlrun-hosts:mlrun:docker-desktop BEGIN\n127.0.0.1 jupyter.mlrun.docker-desktop.lab.iguazeng.com\n127.0.0.1 mlrun-ui.mlrun.docker-desktop.lab.iguazeng.com\n127.0.0.1 mlrun-api.mlrun.docker-desktop.lab.iguazeng.com\n127.0.0.1 nuclio.mlrun.docker-desktop.lab.iguazeng.com\n127.0.0.1 s3.mlrun.docker-desktop.lab.iguazeng.com\n127.0.0.1 minio-console.mlrun.docker-desktop.lab.iguazeng.com\n127.0.0.1 tdengine.mlrun.docker-desktop.lab.iguazeng.com\n# mlrun-hosts:mlrun:docker-desktop END\n" >> /etc/hosts'
```

To remove:

```bash
sudo /bin/sh -c "sed -i '' '/# mlrun-hosts:mlrun:docker-desktop BEGIN/,/# mlrun-hosts:mlrun:docker-desktop END/d' /etc/hosts"
```

### 3) Single namespace (no controller)

Note: `values.yaml` is the chart default. You can omit `-f values.yaml` if you want.

```bash
helm upgrade --install mlrun \
  -n mlrun --create-namespace \
  --insecure-skip-tls-verify \
  ce/charts/mlrun-ce \
  -f ce/charts/mlrun-ce/values.yaml \
  --set mlrun.api.ingress.enabled=true \
  --set mlrun.ui.ingress.enabled=true \
  --set jupyterNotebook.ingress.enabled=true \
  --set minio.ingress.enabled=true \
  --set minio.consoleIngress.enabled=true \
  --set nuclio.dashboard.ingress.enabled=true \
  --set tdengine.ingress.enabled=true \
  --set-string=mlrun.api.ingress.hosts[0].host=mlrun-api.mlrun.docker-desktop.lab.iguazeng.com \
  --set-string=mlrun.ui.ingress.hosts[0].host=mlrun-ui.mlrun.docker-desktop.lab.iguazeng.com \
  --set-string=jupyterNotebook.ingress.hosts[0].host=jupyter.mlrun.docker-desktop.lab.iguazeng.com \
  --set-string=nuclio.dashboard.ingress.hosts[0]=nuclio.mlrun.docker-desktop.lab.iguazeng.com \
  --set-string=minio.ingress.hosts[0]=s3.mlrun.docker-desktop.lab.iguazeng.com \
  --set-string=minio.consoleIngress.hosts[0]=minio-console.mlrun.docker-desktop.lab.iguazeng.com \
  --set-string=tdengine.ingress.hosts[0]=tdengine.mlrun.docker-desktop.lab.iguazeng.com \
  --set-string=mlrun.api.ingress.hosts[0].paths[0].path=/ \
  --set-string=mlrun.api.ingress.hosts[0].paths[0].pathType=Prefix \
  --set-string=mlrun.ui.ingress.hosts[0].paths[0].path=/ \
  --set-string=mlrun.ui.ingress.hosts[0].paths[0].pathType=Prefix \
  --set mlrun.api.ingress.ingressClassName=nginx \
  --set mlrun.ui.ingress.ingressClassName=nginx \
  --set jupyterNotebook.ingress.ingressClassName=nginx \
  --set minio.ingress.ingressClassName=nginx \
  --set minio.consoleIngress.ingressClassName=nginx \
  --set nuclio.dashboard.ingress.ingressClassName=nginx \
  --set tdengine.ingress.ingressClassName=nginx
```

### 4) Multi namespace (controller + per‑namespace)

#### 4.1) Controller (once)

```bash
helm upgrade --install mlrun-ce-controller \
  -n controller --create-namespace \
  --insecure-skip-tls-verify \
  ce/charts/mlrun-ce \
  -f ce/charts/mlrun-ce/admin_installation_values.yaml
```

#### 4.2) Namespace install (repeat per namespace)

Replace `<ns>` and `<cluster>`.

Choose one values file:

- `non_admin_cluster_ip_installation_values.yaml` (ClusterIP)
- `non_admin_installation_values.yaml` (NodePort)

Note: `values.yaml` is the chart default. You can omit `-f values.yaml` if you want.

```bash
helm upgrade --install <ns> \
  -n <ns> --create-namespace \
  --insecure-skip-tls-verify \
  ce/charts/mlrun-ce \
  -f ce/charts/mlrun-ce/non_admin_cluster_ip_installation_values.yaml \
  --set mlrun.api.ingress.enabled=true \
  --set mlrun.ui.ingress.enabled=true \
  --set jupyterNotebook.ingress.enabled=true \
  --set minio.ingress.enabled=true \
  --set minio.consoleIngress.enabled=true \
  --set nuclio.dashboard.ingress.enabled=true \
  --set tdengine.ingress.enabled=true \
  --set-string=mlrun.api.ingress.hosts[0].host=mlrun-api.<ns>.<cluster>.lab.iguazeng.com \
  --set-string=mlrun.ui.ingress.hosts[0].host=mlrun-ui.<ns>.<cluster>.lab.iguazeng.com \
  --set-string=jupyterNotebook.ingress.hosts[0].host=jupyter.<ns>.<cluster>.lab.iguazeng.com \
  --set-string=nuclio.dashboard.ingress.hosts[0]=nuclio.<ns>.<cluster>.lab.iguazeng.com \
  --set-string=minio.ingress.hosts[0]=s3.<ns>.<cluster>.lab.iguazeng.com \
  --set-string=minio.consoleIngress.hosts[0]=minio-console.<ns>.<cluster>.lab.iguazeng.com \
  --set-string=tdengine.ingress.hosts[0]=tdengine.<ns>.<cluster>.lab.iguazeng.com \
  --set-string=mlrun.api.ingress.hosts[0].paths[0].path=/ \
  --set-string=mlrun.api.ingress.hosts[0].paths[0].pathType=Prefix \
  --set-string=mlrun.ui.ingress.hosts[0].paths[0].path=/ \
  --set-string=mlrun.ui.ingress.hosts[0].paths[0].pathType=Prefix \
  --set mlrun.api.ingress.ingressClassName=nginx \
  --set mlrun.ui.ingress.ingressClassName=nginx \
  --set jupyterNotebook.ingress.ingressClassName=nginx \
  --set minio.ingress.ingressClassName=nginx \
  --set minio.consoleIngress.ingressClassName=nginx \
  --set nuclio.dashboard.ingress.ingressClassName=nginx \
  --set tdengine.ingress.ingressClassName=nginx
```

#### NodePort variant (manual)

If you prefer NodePort instead of ClusterIP, replace the ClusterIP values file:

```bash
helm upgrade --install <ns> \
  -n <ns> --create-namespace \
  --insecure-skip-tls-verify \
  ce/charts/mlrun-ce \

  -f ce/charts/mlrun-ce/non_admin_installation_values.yaml \
  --set mlrun.api.ingress.enabled=true \
  --set mlrun.ui.ingress.enabled=true \
  --set jupyterNotebook.ingress.enabled=true \
  --set minio.ingress.enabled=true \
  --set minio.consoleIngress.enabled=true \
  --set nuclio.dashboard.ingress.enabled=true \
  --set tdengine.ingress.enabled=true \
  --set-string=mlrun.api.ingress.hosts[0].host=mlrun-api.<ns>.<cluster>.lab.iguazeng.com \
  --set-string=mlrun.ui.ingress.hosts[0].host=mlrun-ui.<ns>.<cluster>.lab.iguazeng.com \
  --set-string=jupyterNotebook.ingress.hosts[0].host=jupyter.<ns>.<cluster>.lab.iguazeng.com \
  --set-string=nuclio.dashboard.ingress.hosts[0]=nuclio.<ns>.<cluster>.lab.iguazeng.com \
  --set-string=minio.ingress.hosts[0]=s3.<ns>.<cluster>.lab.iguazeng.com \
  --set-string=minio.consoleIngress.hosts[0]=minio-console.<ns>.<cluster>.lab.iguazeng.com \
  --set-string=tdengine.ingress.hosts[0]=tdengine.<ns>.<cluster>.lab.iguazeng.com \
  --set-string=mlrun.api.ingress.hosts[0].paths[0].path=/ \
  --set-string=mlrun.api.ingress.hosts[0].paths[0].pathType=Prefix \
  --set-string=mlrun.ui.ingress.hosts[0].paths[0].path=/ \
  --set-string=mlrun.ui.ingress.hosts[0].paths[0].pathType=Prefix \
  --set mlrun.api.ingress.ingressClassName=nginx \
  --set mlrun.ui.ingress.ingressClassName=nginx \
  --set jupyterNotebook.ingress.ingressClassName=nginx \
  --set minio.ingress.ingressClassName=nginx \
  --set minio.consoleIngress.ingressClassName=nginx \
  --set nuclio.dashboard.ingress.ingressClassName=nginx \
  --set tdengine.ingress.ingressClassName=nginx
```

---

## Verify

```bash
kubectl get pods -n <ns>
kubectl get ing -n <ns>
```

GiladShapira94 commented on charts/mlrun-ce/values.yaml:
```diff
@@ -489,37 +489,73 @@ timescaledb:
   nodeSelector: {}
   tolerations: []
 
+strimzi-kafka-operator:
```
Make sure that there are no redundant values based on the https://strimzi.io/charts/ values file

shay79il replied:
Done

shay79il committed:
Add Helm hooks to Kafka CRs to ensure CRDs are installed first

- Add post-install/post-upgrade hooks to kafka-cluster.yaml
- Add post-install/post-upgrade hooks to kafka-nodepool.yaml
- Set hook-weight to 5 to ensure they run after operator installation

This fixes the 'resource mapping not found' error when deploying
Kafka CRs before the Strimzi CRDs are installed.
